pkg/search/controller.go

```diff
@@ -161,46 +163,210 @@ func (c *Controller) handleErr(err error, key interface{}) {
-	v, ok := c.clusterRegistry.Load(cluster)
-	if !ok {
-		klog.Infof("Cluster %s is not registered", cluster)
-		return nil
+func (c *Controller) getClusterMatchedRegistries(cluster *clusterv1alpha1.Cluster) (indexedByName, matched map[string]*searchv1alpha1.ResourceRegistry, err error) {
+	client := c.informerFactory.Search().V1alpha1().ResourceRegistries().Lister()
+	var registries []*searchv1alpha1.ResourceRegistry
+	if registries, err = client.List(labels.Everything()); err != nil {
+		klog.Errorf("List resource registries for reconciling cluster failed, err: %s", err)
+		return
+	}
+	if len(registries) == 0 {
+		klog.Infof("No resource registries, no need to reconcile cluster")
+		return
 	}
+	indexedByName = make(map[string]*searchv1alpha1.ResourceRegistry, len(registries))
+	matched = make(map[string]*searchv1alpha1.ResourceRegistry, len(registries))
+	for i, registry := range registries {
+		indexedByName[registry.Name] = registries[i]
+		if util.ClusterMatches(cluster, registry.Spec.TargetCluster) {
+			matched[registry.Name] = registries[i]
+		}
+	}
+	return
+}
 
-	cr := v.(clusterRegistry)
-	if cr.unregistry() {
-		klog.Infof("try to stop cluster informer %s", cluster)
-		c.InformerManager.Stop(cluster)
-		return nil
+func (c *Controller) getRegistryAddedResources(registry *searchv1alpha1.ResourceRegistry, cr *clusterRegistry, added []schema.GroupVersionResource) []schema.GroupVersionResource {
+	name := registry.Name
+	resourcesToWatch := c.getResources(registry.Spec.ResourceSelectors)
+	for _, resource := range resourcesToWatch {
+		if resourcesOnWatching, exist := cr.resources[resource]; exist {
+			resourcesOnWatching[name] = struct{}{}
+		} else {
+			added = append(added, resource)
+			cr.resources[resource] = registrySet{name: struct{}{}}
+		}
+	}
+	return added
+}
+
+func (c *Controller) getClusterRegistriesModification(registries map[string]*searchv1alpha1.ResourceRegistry, matched map[string]*searchv1alpha1.ResourceRegistry, cr *clusterRegistry, added []string, removed []string) (updated *clusterRegistry, addedResources []schema.GroupVersionResource, removedResources []schema.GroupVersionResource) {
+	defer func() {
+		updated = cr
+	}()
+	if cr == nil {
+		cr = &clusterRegistry{
+			resources:  make(map[schema.GroupVersionResource]registrySet),
+			registries: make(registrySet)}
+	}
+	for _, name := range added {
+		cr.registries[name] = struct{}{}
+		addedResources = c.getRegistryAddedResources(registries[name], cr, addedResources)
+	}
+	for _, name := range removed {
+		delete(cr.registries, name)
+		for resource, previousRegistries := range cr.resources {
+			delete(previousRegistries, name)
+			if len(previousRegistries) == 0 {
+				removedResources = append(removedResources, resource)
+			}
+		}
+		for _, resource := range removedResources {
+			delete(cr.resources, resource)
+		}
+	}
+	for _, registry := range matched {
+		addedResources = c.getRegistryAddedResources(registry, cr, addedResources)
+	}
```

Comment:
This processing step feels redundant. All the matched RRs should have already been processed.

Reply:
No, there may be added ones, and removed ones. But still remaining ones may have resources changed.
Like, registry target cluster is `aaa=bbb`, resources, `Pod`, `Node`, cluster label changed, and resources also changed, like adding `Event`.
So we not only need to handle registries added or removed, but also need to handle resources changes.